src/codemagic/tools/app_store_connect/actions/fetch_signing_files_action.py

```diff
@@ -0,0 +1,237 @@
+                certificate_key=certificate_key,
+                certificate_key_password=certificate_key_password,
+                should_print=False,
+            )
+            certificates.append(certificate)
+        return certificates
+
+    def _get_or_create_profiles(
+        self,
+        bundle_ids: Sequence[BundleId],
+        certificates: Sequence[SigningCertificate],
+        profile_type: ProfileType,
+        create_resource: bool,
+        platform: Optional[BundleIdPlatform] = None,
+    ):
+        def has_certificate(profile) -> bool:
+            try:
+                profile_certificates = self.api_client.profiles.list_certificate_ids(profile)
+                return bool(certificate_ids.issubset({c.id for c in profile_certificates}))
+            except AppStoreConnectApiError as err:
+                error = f"Listing {SigningCertificate.s} for {Profile} {profile.id} failed unexpectedly"
+                self.logger.warning(Colors.YELLOW(f"{error}: {err.error_response}"))
+                return False
+
+        def missing_profile(bundle_id) -> bool:
+            try:
+                bundle_ids_profiles = self.api_client.bundle_ids.list_profile_ids(bundle_id)
+                return not (profile_ids & {p.id for p in bundle_ids_profiles})
+            except AppStoreConnectApiError as err:
+                error = f"Listing {Profile.s} for {BundleId} {bundle_id.id} failed unexpectedly"
+                self.logger.warning(Colors.YELLOW(f"{error}: {err.error_response}"))
+                return True
+
+        certificate_ids = {c.id for c in certificates}
+        profiles = self.list_bundle_id_profiles(
+            [bundle_id.id for bundle_id in bundle_ids],
+            profile_type=profile_type,
+            profile_state=ProfileState.ACTIVE,
+            should_print=False,
+        )
+        profiles = list(filter(has_certificate, profiles))
+
+        certificate_names = ", ".join(c.get_display_info() for c in certificates)
+        message = f"that contain {SigningCertificate.plural(len(certificates))} {certificate_names}"
+        self.printer.log_filtered(Profile, profiles, message)
+        for profile in profiles:
+            self.logger.info(f"- {profile.get_display_info()}")
+
+        profile_ids = {p.id for p in profiles}
+        bundle_ids_without_profiles = list(filter(missing_profile, bundle_ids))
+        if bundle_ids_without_profiles and not create_resource:
+            missing = ", ".join(f'"{bid.attributes.identifier}" [{bid.id}]' for bid in bundle_ids_without_profiles)
+            raise AppStoreConnectError(f"Did not find {profile_type} {Profile.s} for {BundleId.s}: {missing}")
+
+        created_profiles = self._create_missing_profiles(
+            bundle_ids_without_profiles,
+            certificates,
+            profile_type,
+            platform,
+        )
+        profiles.extend(created_profiles)
+        return profiles
+
+    def _create_missing_profiles(
```

Comment:
Used to be defined in https://github.com/codemagic-ci-cd/cli-tools/blob/3670fa624997f17cb437bc88198f7bdee3594f35/src/codemagic/tools/app_store_connect.py#L947-L970